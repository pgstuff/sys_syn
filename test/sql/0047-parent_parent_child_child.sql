BEGIN;

CREATE EXTENSION tinyint SCHEMA public;
CREATE EXTENSION sys_syn;

CREATE SCHEMA user_data
    AUTHORIZATION postgres;

CREATE TABLE user_data.parent_1 (
        parent_1_id integer NOT NULL,
        parent_1_text text,
        CONSTRAINT parent_1_pid PRIMARY KEY (parent_1_id));

CREATE TABLE user_data.parent_2 (
        parent_2_id integer NOT NULL,
        parent_2_text text,
        CONSTRAINT parent_2_pid PRIMARY KEY (parent_2_id));

CREATE TABLE user_data.child_1 (
        child_1_id integer NOT NULL,
        parent_1_id integer,
        CONSTRAINT child_1_pid PRIMARY KEY (child_1_id));

CREATE TABLE user_data.child_2 (
        child_2_id integer NOT NULL,
        parent_1_id integer,
        parent_2_id integer,
        CONSTRAINT child_2_pid PRIMARY KEY (child_2_id));

INSERT INTO sys_syn.in_groups_def VALUES ('in');

DO $$BEGIN
        EXECUTE sys_syn.in_table_create_sql('user_data.parent_1'::regclass, 'in');
        EXECUTE sys_syn.in_table_create_sql('user_data.parent_2'::regclass, 'in');
END$$;

SELECT sys_syn.in_table_create (
                schema          => 'user_data',
                in_table_id     => 'child_1',
                in_group_id     => 'in',
                in_pull_id      => NULL,
                in_columns      => ARRAY[
                       $COL$("child_1_id","integer",Id,"in_source.child_1_id",,,,,)$COL$,
                       $COL$("parent_1_id","integer",Attribute,"in_source.parent_1_id",,1,"parent_1","parent_1_id",)$COL$
                ]::sys_syn.create_in_column[],
                full_table_reference    => 'user_data.child_1'
        );

SELECT sys_syn.in_table_create (
                schema          => 'user_data',
                in_table_id     => 'child_2',
                in_group_id     => 'in',
                in_pull_id      => NULL,
                in_columns      => ARRAY[
                       $COL$("child_2_id","integer",Id,"in_source.child_2_id",,,,,)$COL$,
                       $COL$("parent_1_id","integer",Attribute,"in_source.parent_1_id",,1,"parent_1","parent_1_id",)$COL$,
                       $COL$("parent_2_id","integer",Attribute,"in_source.parent_2_id",,2,"parent_2","parent_2_id",)$COL$
                ]::sys_syn.create_in_column[],
                full_table_reference    => 'user_data.child_2'
        );

INSERT INTO user_data.parent_1 (
        parent_1_id,   parent_1_text)
VALUES (1,              'parent_1_data');

INSERT INTO user_data.parent_2 (
        parent_2_id,   parent_2_text)
VALUES (2,              'parent_2_data');

INSERT INTO user_data.child_1 (
        child_1_id,    parent_1_id)
VALUES (3,              1);

INSERT INTO user_data.child_2 (
        child_2_id,    parent_1_id,   parent_2_id)
VALUES (4,              1,              2);

INSERT INTO sys_syn.out_groups_def VALUES ('out');

SELECT sys_syn.out_table_create('user_data', 'parent_1', 'out');

SELECT sys_syn.out_table_create('user_data', 'parent_2', 'out');

SELECT sys_syn.out_table_create('user_data', 'child_1', 'out');

SELECT sys_syn.out_table_create('user_data', 'child_2', 'out');

ALTER TABLE user_data.parent_1_out_queue_1
  ADD FOREIGN KEY (trans_id_in, id) REFERENCES user_data.parent_1_in_1 (trans_id_in, id) ON UPDATE RESTRICT ON DELETE RESTRICT;

SELECT user_data.parent_1_pull(FALSE);
SELECT user_data.parent_1_out_move_1();
SELECT id, delta_type, queue_state FROM user_data.parent_1_out_queue_1;

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;SET LOCAL sys_syn.trans_id_curr TO 2;
SELECT user_data.child_1_pull(FALSE);
SELECT user_data.child_1_out_move_1();
SELECT id, delta_type, queue_state FROM user_data.child_1_out_queue_1;

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;SET LOCAL sys_syn.trans_id_curr TO 3;
SELECT user_data.child_2_pull(FALSE);
SELECT user_data.child_2_out_move_1();
SELECT id, delta_type, queue_state FROM user_data.child_2_out_queue_1;

UPDATE user_data.parent_1_out_queue_1 SET queue_state = 'Claimed'::sys_syn.queue_state, hold_reason_text = 'Testing parent-child dependency.' WHERE (id).parent_1_id = 1;
UPDATE user_data.parent_1_out_queue_1 SET queue_state = 'Hold'::sys_syn.queue_state, hold_reason_text = 'Testing parent-child dependency.' WHERE (id).parent_1_id = 1;
SELECT user_data.parent_1_out_processed_1();
SELECT id, delta_type, queue_state FROM user_data.parent_1_out_queue_1;

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;SET LOCAL sys_syn.trans_id_curr TO 4;
SELECT user_data.child_1_pull(FALSE);
SELECT user_data.child_1_out_move_1();
SELECT id, delta_type, queue_state FROM user_data.child_1_out_queue_1;

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;SET LOCAL sys_syn.trans_id_curr TO 5;
SELECT user_data.child_2_pull(FALSE);
SELECT user_data.child_2_out_move_1();
SELECT id, delta_type, queue_state FROM user_data.child_2_out_queue_1;

UPDATE user_data.parent_1_out_queue_1 SET queue_state = 'Claimed'::sys_syn.queue_state WHERE (id).parent_1_id = 1;
UPDATE user_data.parent_1_out_queue_1 SET queue_state = 'Processed'::sys_syn.queue_state WHERE (id).parent_1_id = 1;
SELECT user_data.parent_1_out_processed_1();
SELECT id, delta_type, queue_state FROM user_data.parent_1_out_queue_1;

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;SET LOCAL sys_syn.trans_id_curr TO 6;
SELECT user_data.child_1_pull(FALSE);
SELECT user_data.child_1_out_move_1();
SELECT id, delta_type, queue_state FROM user_data.child_1_out_queue_1;

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;SET LOCAL sys_syn.trans_id_curr TO 7;
SELECT user_data.child_2_pull(FALSE);
SELECT user_data.child_2_out_move_1();
SELECT id, delta_type, queue_state FROM user_data.child_2_out_queue_1;

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;SET LOCAL sys_syn.trans_id_curr TO 8;
SELECT user_data.parent_2_pull(FALSE);
SELECT user_data.parent_2_out_move_1();
SELECT id, delta_type, queue_state FROM user_data.parent_2_out_queue_1;

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;SET LOCAL sys_syn.trans_id_curr TO 9;
SELECT user_data.child_2_pull(FALSE);
SELECT user_data.child_2_out_move_1();
SELECT id, delta_type, queue_state FROM user_data.child_2_out_queue_1;

UPDATE user_data.parent_2_out_queue_1 SET queue_state = 'Claimed'::sys_syn.queue_state WHERE (id).parent_2_id = 2;
UPDATE user_data.parent_2_out_queue_1 SET queue_state = 'Processed'::sys_syn.queue_state WHERE (id).parent_2_id = 2;
SELECT user_data.parent_2_out_processed_1();
SELECT id, delta_type, queue_state FROM user_data.parent_2_out_queue_1;

SELECT user_data.child_2_out_foreign_processed_1();
SELECT id, delta_type, queue_state FROM user_data.child_2_out_queue_1;

ROLLBACK;
