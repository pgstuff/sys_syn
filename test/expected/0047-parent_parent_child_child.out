BEGIN;
CREATE EXTENSION tinyint
    SCHEMA public;
CREATE EXTENSION sys_syn;
CREATE SCHEMA user_data
    AUTHORIZATION postgres;
CREATE TABLE user_data.parent_1 (
        parent_1_key integer NOT NULL,
        parent_1_text text,
        CONSTRAINT parent_1_pkey PRIMARY KEY (parent_1_key));
CREATE TABLE user_data.parent_2 (
        parent_2_key integer NOT NULL,
        parent_2_text text,
        CONSTRAINT parent_2_pkey PRIMARY KEY (parent_2_key));
CREATE TABLE user_data.child_1 (
        child_1_key integer NOT NULL,
        parent_1_key integer,
        CONSTRAINT child_1_pkey PRIMARY KEY (child_1_key));
CREATE TABLE user_data.child_2 (
        child_2_key integer NOT NULL,
        parent_1_key integer,
        parent_2_key integer,
        CONSTRAINT child_2_pkey PRIMARY KEY (child_2_key));
INSERT INTO sys_syn.in_groups_def VALUES ('in');
DO $$BEGIN
        EXECUTE sys_syn.in_table_add_sql('user_data.parent_1'::regclass, 'in');
        EXECUTE sys_syn.in_table_add_sql('user_data.parent_2'::regclass, 'in');
END$$;
SELECT sys_syn.in_table_add (
                schema          => 'user_data',
                in_table_id     => 'child_1',
                in_group_id     => 'in',
                in_pull_id      => NULL,
                in_columns      => ARRAY[
                       $COL$("child_1_key","integer",Key,"in_source.child_1_key",,,,)$COL$,
                       $COL$("parent_1_key","integer",Attribute,"in_source.parent_1_key",,1,"parent_1","parent_1_key")$COL$
                ]::sys_syn.create_in_column[],
                full_table_reference    => 'user_data.child_1'
        );
 in_table_add 
--------------
 
(1 row)

SELECT sys_syn.in_table_add (
                schema          => 'user_data',
                in_table_id     => 'child_2',
                in_group_id     => 'in',
                in_pull_id      => NULL,
                in_columns      => ARRAY[
                       $COL$("child_2_key","integer",Key,"in_source.child_2_key",,,,)$COL$,
                       $COL$("parent_1_key","integer",Attribute,"in_source.parent_1_key",,1,"parent_1","parent_1_key")$COL$,
                       $COL$("parent_2_key","integer",Attribute,"in_source.parent_2_key",,2,"parent_2","parent_2_key")$COL$
                ]::sys_syn.create_in_column[],
                full_table_reference    => 'user_data.child_2'
        );
 in_table_add 
--------------
 
(1 row)

INSERT INTO user_data.parent_1 (
        parent_1_key,   parent_1_text)
VALUES (1,              'parent_1_data');
INSERT INTO user_data.parent_2 (
        parent_2_key,   parent_2_text)
VALUES (2,              'parent_2_data');
INSERT INTO user_data.child_1 (
        child_1_key,    parent_1_key)
VALUES (3,              1);
INSERT INTO user_data.child_2 (
        child_2_key,    parent_1_key,   parent_2_key)
VALUES (4,              1,              2);
INSERT INTO sys_syn.out_groups_def VALUES ('out');
SELECT sys_syn.out_table_add('user_data', 'parent_1', 'out');
 out_table_add 
---------------
 
(1 row)

SELECT sys_syn.out_table_add('user_data', 'parent_2', 'out');
 out_table_add 
---------------
 
(1 row)

SELECT sys_syn.out_table_add('user_data', 'child_1', 'out');
 out_table_add 
---------------
 
(1 row)

SELECT sys_syn.out_table_add('user_data', 'child_2', 'out');
 out_table_add 
---------------
 
(1 row)

ALTER TABLE user_data.parent_1_out_queue
  ADD FOREIGN KEY (trans_id_in, key) REFERENCES user_data.parent_1_in (trans_id_in, key) ON UPDATE RESTRICT ON DELETE RESTRICT;
SELECT user_data.parent_1_pull(FALSE);
 parent_1_pull 
---------------
 t
(1 row)

SELECT user_data.parent_1_out_move();
 parent_1_out_move 
-------------------
 t
(1 row)

SELECT key, delta_type, queue_state FROM user_data.parent_1_out_queue;
 key | delta_type | queue_state 
-----+------------+-------------
 (1) | Add        | Unread
(1 row)

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;
SELECT user_data.child_1_pull(FALSE);
 child_1_pull 
--------------
 t
(1 row)

SELECT user_data.child_1_out_move();
 child_1_out_move 
------------------
 f
(1 row)

SELECT key, delta_type, queue_state FROM user_data.child_1_out_queue;
 key | delta_type | queue_state 
-----+------------+-------------
(0 rows)

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;
SELECT user_data.child_2_pull(FALSE);
 child_2_pull 
--------------
 t
(1 row)

SELECT user_data.child_2_out_move();
 child_2_out_move 
------------------
 f
(1 row)

SELECT key, delta_type, queue_state FROM user_data.child_2_out_queue;
 key | delta_type | queue_state 
-----+------------+-------------
(0 rows)

UPDATE user_data.parent_1_out_queue SET queue_state = 'Hold'::sys_syn.queue_state WHERE (key).parent_1_key = 1;
SELECT user_data.parent_1_out_processed();
 parent_1_out_processed 
------------------------
 f
(1 row)

SELECT key, delta_type, queue_state FROM user_data.parent_1_out_queue;
 key | delta_type | queue_state 
-----+------------+-------------
 (1) | Add        | Hold
(1 row)

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;
SELECT user_data.child_1_pull(FALSE);
 child_1_pull 
--------------
 t
(1 row)

SELECT user_data.child_1_out_move();
 child_1_out_move 
------------------
 f
(1 row)

SELECT key, delta_type, queue_state FROM user_data.child_1_out_queue;
 key | delta_type | queue_state 
-----+------------+-------------
(0 rows)

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;
SELECT user_data.child_2_pull(FALSE);
 child_2_pull 
--------------
 t
(1 row)

SELECT user_data.child_2_out_move();
 child_2_out_move 
------------------
 f
(1 row)

SELECT key, delta_type, queue_state FROM user_data.child_2_out_queue;
 key | delta_type | queue_state 
-----+------------+-------------
(0 rows)

UPDATE user_data.parent_1_out_queue SET queue_state = 'Processed'::sys_syn.queue_state WHERE (key).parent_1_key = 1;
SELECT user_data.parent_1_out_processed();
 parent_1_out_processed 
------------------------
 t
(1 row)

SELECT key, delta_type, queue_state FROM user_data.parent_1_out_queue;
 key | delta_type | queue_state 
-----+------------+-------------
(0 rows)

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;
SELECT user_data.child_1_pull(FALSE);
 child_1_pull 
--------------
 t
(1 row)

SELECT user_data.child_1_out_move();
 child_1_out_move 
------------------
 t
(1 row)

SELECT key, delta_type, queue_state FROM user_data.child_1_out_queue;
 key | delta_type | queue_state 
-----+------------+-------------
 (3) | Add        | Unread
(1 row)

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;
SELECT user_data.child_2_pull(FALSE);
 child_2_pull 
--------------
 t
(1 row)

SELECT user_data.child_2_out_move();
 child_2_out_move 
------------------
 f
(1 row)

SELECT key, delta_type, queue_state FROM user_data.child_2_out_queue;
 key | delta_type | queue_state 
-----+------------+-------------
(0 rows)

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;
SELECT user_data.parent_2_pull(FALSE);
 parent_2_pull 
---------------
 t
(1 row)

SELECT user_data.parent_2_out_move();
 parent_2_out_move 
-------------------
 t
(1 row)

SELECT key, delta_type, queue_state FROM user_data.parent_2_out_queue;
 key | delta_type | queue_state 
-----+------------+-------------
 (2) | Add        | Unread
(1 row)

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;
SELECT user_data.child_2_pull(FALSE);
 child_2_pull 
--------------
 t
(1 row)

SELECT user_data.child_2_out_move();
 child_2_out_move 
------------------
 f
(1 row)

SELECT key, delta_type, queue_state FROM user_data.child_2_out_queue;
 key | delta_type | queue_state 
-----+------------+-------------
(0 rows)

UPDATE user_data.parent_2_out_queue SET queue_state = 'Processed'::sys_syn.queue_state WHERE (key).parent_2_key = 2;
SELECT user_data.parent_2_out_processed();
 parent_2_out_processed 
------------------------
 t
(1 row)

SELECT key, delta_type, queue_state FROM user_data.parent_2_out_queue;
 key | delta_type | queue_state 
-----+------------+-------------
(0 rows)

SELECT user_data.child_2_out_foreign_processed();
 child_2_out_foreign_processed 
-------------------------------
 t
(1 row)

SELECT key, delta_type, queue_state FROM user_data.child_2_out_queue;
 key | delta_type | queue_state 
-----+------------+-------------
 (4) | Add        | Unread
(1 row)

ROLLBACK;
